mlir/include/mlir/Dialect/DLTI/DLTIAttrs.td

```diff
@@ -0,0 +1,303 @@
+}
+
+//===----------------------------------------------------------------------===//
+// TargetSystemDescSpecAttr
+//===----------------------------------------------------------------------===//
+
+def TargetSystemDescSpecTrait
+    : NativeAttrTrait<"TargetSystemDescSpecInterface::Trait"> {
+  let cppNamespace = "::mlir";
+}
+
+def DLTI_TargetSystemDescSpecAttr :
+    DLTIAttr<"TargetSystemDescSpec", [TargetSystemDescSpecTrait]> {
+  let summary = [{An attribute to represent target system description}];
+  let description = [{
+    A system description specification describes the overall system
+    containing multiple devices, with each device having a unique ID
+    and its corresponding TargetDeviceDescSpec object.
+
+    Example:
+          dlti.target_system_desc_spec =
+            #dlti.target_device_desc_spec<
+              #dlti.dl_entry<"dlti.device_id", 0: ui32>,
+              #dlti.dl_entry<"dlti.device_type", "CPU">>,
+            #dlti.target_device_desc_spec <
+              #dlti.dl_entry<"dlti.device_id", 1: ui32>,
+              #dlti.dl_entry<"dlti.device_type", "GPU">,
+              #dlti.dl_entry<"dlti.max_vector_op_width", 64 : ui32>>,
+            #dlti.target_device_desc_spec <
+              #dlti.dl_entry<"dlti.device_id", 2: ui32>,
+              #dlti.dl_entry<"dlti.device_type", "XPU">>>
+  }];
+  let parameters = (ins
+    ArrayRefParameter<"TargetDeviceDescSpecInterface", "">:$entries
+  );
+  let mnemonic = "target_system_desc_spec";
+  let genVerifyDecl = 1;
+  let hasCustomAssemblyFormat = 1;
+  let extraClassDeclaration = [{
+    /// Return the device descriptor that matches the given device ID
+    std::optional<TargetDeviceDescSpecInterface>
+    getDeviceDescForDeviceID(uint32_t deviceID);
+  }];
+  let extraClassDefinition = [{
+    std::optional<TargetDeviceDescSpecInterface>
+    $cppClass::getDeviceDescForDeviceID(
+        TargetDeviceDescSpecInterface::DeviceID deviceID) {
+      for (TargetDeviceDescSpecInterface entry : getEntries()) {
+        if (entry.getDeviceID() == deviceID)
+          return entry;
+      }
+      return std::nullopt;
+    }
+  }];
+}
+
+//===----------------------------------------------------------------------===//
+// TargetDeviceDescSpecAttr
+//===----------------------------------------------------------------------===//
+
+def TargetDeviceDescSpecTrait
+    : NativeAttrTrait<"TargetDeviceDescSpecInterface::Trait"> {
+  let cppNamespace = "::mlir";
+}
+
+def DLTI_TargetDeviceDescSpecAttr :
+    DLTIAttr<"TargetDeviceDescSpec", [TargetDeviceDescSpecTrait]> {
+  let summary = [{An attribute to represent target device description}];
+  let description = [{
+        Each device description specification describes a single device and
+        its hardware properties. Each device description must have a device_id
+        and a device_type. In addition, the description can contain any number
+        of optional hardware properties (e.g., max_vector_op_width below).
```

Comment:
Shouldn't optional properties be marked as e.g. `#dlti.dl_optional_entry` rather than `#dlti.dl_entry`? That would make things easier for humans.

Reply:
That's an interesting thought. So far, the PR only mandates `device_id` and `device_type` properties. Rest all are optional. As the distinction is clear so far, I did not feel the need to create a separate data layout entry type. But if we all feel the need (or as the number of supported mandatory properties grows), we can create an optional property.